### ISSUE TITLE
Embeddings classification ability

### ISSUE THREAD

**Issue by lodemo** [davidsandberg/facenet#134]
Hello,
i wanted to create a comparison between different face recognition techniques in a classification context.
For this i reused the classification experiment from Openface (https://github.com/cmusatyalab/openface/blob/master/evaluation/lfw-classification.py) and integrated Facenet into it.
The results are however not as i would have thought. Even at 10 different people accuracy is only around 0.5. Are the resulting image embeddings different to the one from Openface in their ability to be used directly in classification, using a classifier like SVM?

At the moment im feeding a single image to the network like this and use the resulting embedding as the representation for classification/evaluation:

`imgs = np.reshape(img, (1, 160, 160, 3)) # img is (160, 160, 3)`
`feed_dict = { images_placeholder:imgs }`
`emb = sess.run(embeddings, feed_dict=feed_dict)`
`rep = emb[0]`

Or is the error perhaps in the evaluation of the results? Currently using the same accuracy calculation like Openface, the accuracy_score function from sklearn.metrics.

regards

**Comment by davidsandberg** [comment]
The loss used for training is different in openface compared to this implementation, but I don't think that should cause any major differences in how the resulting embedding can be used.
Could it be something related to scaling/normalization of the input image?

**Comment by lodemo** [comment]
Thank you for the suggestion,
i use the Mtcnn aligned LFW dataset with size 160, like described in Validate on LFW.
In the Openface script, cv2 is used reading the data with the following,

` img = cv2.imread(imgPath)`
` img = cv2.resize(img, (size, size))`
`...`
`img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)`

which seems to load the image with cv2.imread as BGR and converts it to RGB.
RGB is correct for facenet? Size should not be changed as i specified 160.
I will further look into it anyway.

**Comment by lodemo** [comment]
I did further look into it and replaced the data reading with the routine used in Facenet.
I also missed prewhitening completely which probably was the cause of it, or scipy misc.imread() does return different data than cv2.imread()...

After this the results did improve significantly, even more than i expected, compared to the Openface results.

![accuracies](https://cloud.githubusercontent.com/assets/2711650/22370480/e1bc9ff4-e491-11e6-95a3-66a9827a2644.png)

What would you say is the cause for such a difference to Openface, the different loss function? as the embedding dimensions are the same.

**Comment by davidsandberg** [comment]
Hi @lodemo!
That's a very nice result! It would be nice to see how many distractors are needed before the accuracy for "Facenet LinearLVC" starts to degrade. 
For the differences I guess there are a few. I have never managed to get training using triplet loss to work very well. Training the model as a classifier seems to be much easier but it becomes impossible if the number of classes is too large. Also, I never managed to get the NN4 model to converge, so a solution to that was to use the Inception-Resnet models instead. And finally the alignment of training images is different and the MTCNN is very robust to partial occlusion, siluettes etc which makes the training data more diverse.
Which model did you use in this experiment? I have just uploaded a model trained on Ms-Celeb-1M that performs 4 times better than the 20161116-234200 model in terms of LFW errors.

**Comment by lodemo** [comment]
Hello,
i used the 20161116-234200 model from the link in Validate on LFW, but i will try it again with the new model!

What do you mean exactly under distractors? Higher number of people? plan to do that in the next runs.

**Comment by lodemo** [comment]
Hi,
i tried running it with the 20170117-215115 model but encountered an error related to the 
model-20170117-215115.ckpt-285000.data-00000-of-00001 file.

Error:
`tensorflow/core/framework/op_kernel.cc:975] Data loss: Unable to open table file 20170117-215115/model-20170117-215115.ckpt-285000.data-00000-of-00001: Data loss: not an sstable (bad magic number): perhaps your file is in a different file format and you need to use a different restore operator?`

Is the file corrupt or does it need to be loaded differently as the 20161116-234200 model?

**Comment by davidsandberg** [comment]
Hi,
The new restore adds the last part to the file name so the only first part that should be given, i.e. 
`20170117-215115/model-20170117-215115.ckpt-285000` in this case.

**Comment by lodemo** [comment]
Thank you, 
it seems to work, i will run it and post results later.

regards

**Comment by lodemo** [comment]
I ran the scripts with the new model 20170117-215115, it seems to improved the result further.

More prominently in the higher number of people, which i increased up to 1600 (the number of people in LFW with 2 or more images)

Model 20161116-234200:
`400 people: 0.992176 accuracy`
`1600 people: 0.864177 accuracy`

Model 20170117-215115: 
`400 people: 0.997066  accuracy`
`1600 people: 0.926279 accuracy`

![accuracies](https://cloud.githubusercontent.com/assets/2711650/22487753/ec7e0ee8-e80e-11e6-8d69-9aebec5064d0.png)

![accuracies](https://cloud.githubusercontent.com/assets/2711650/22487773/053fb9a4-e80f-11e6-91a3-1211f8750000.png)

**Comment by hardfish82** [comment]
@lodemo It's amazing work! I follow [https://github.com/cmusatyalab/openface/blob/master/evaluation/lfw-classification.py](lfw-classification.py) and add the facenet model(20170117-215115), my result is not so good as you got. What did you improve? The C parameter of LinearSVC?

![accuracies](https://cloud.githubusercontent.com/assets/3353361/22776093/52b828c0-eee9-11e6-8735-12ef6694a612.png)

**Comment by lodemo** [comment]
Actually not, i did leave C=1 but i used the One vs. All scheme explicitly
`LinearSVC(C=1, multi_class='ovr')`

Did you apply all the preprocessing for the Facenet images?
RGB convert, Crop and Prewhiting of the Mtcnn aligned images is important.

I am currently also running evaluation of Openface and Facenet on the YouTube faces database.
Will post results when finished.

**Comment by hardfish82** [comment]
I used SVC API like this:
`cls = SVC(kernel='linear', C=1)`
And just now, I test your scheme that:
`cls = LinearSVC(C=1, multi_class='ovr')`
The result improves slightly:
![accuracies](https://cloud.githubusercontent.com/assets/3353361/22777417/1adf7890-eeee-11e6-9d80-ec0495e5cbc5.png)
I did data preprocessing according facenet code, in the function getData() I added a new mode 'facenet' and used the facenet api to load data:

```
def getData(lfwPpl, nPpl, nImgs, mode, resize=96):
    X, y = [], []

    personNum = 0
    for (person, nTotalImgs) in lfwPpl[:nPpl]:
        imgs = sorted(os.listdir(person))
        for imgPath in imgs[:nImgs]:
            imgPath = os.path.join(person, imgPath)
            img = cv2.imread(imgPath)
            img = cv2.resize(img, (resize, resize))
            if mode == 'grayscale':
                img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
            elif mode == 'rgb':
                img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
            elif mode == 'facenet':
                img = facenet.load_data([imgPath], False, False, resize)[0]
            else:
                assert 0

            X.append(img)
            y.append(personNum)

        personNum += 1

    X = np.array(X)
    y = np.array(y)
    return (X, y)
```

**Comment by lodemo** [comment]
I see, its probably because you resize the images for facenet to 96, which if you use the pre-trained model should be 160.

**Comment by hardfish82** [comment]
Yes, The lfw images are aligned to size 160*160 and I got facenet data with resize=160. Otherwise a shape mismatch error would be raised.
```
    net = facenet_model.FaceNet("/home/x0269/models/20170131-234652")
    facenetGPUsvmDf = cacheToFile(cache)(openfaceExp)(lfwPpl, net, cls, 'facenet', 160)
```
Did you filter lfw images with [https://github.com/cmusatyalab/openface/blob/master/util/prune-dataset.py](prune-dataset)? I did it with --numImagesThreshold=5, then the persons with less than 5 images are filtered. The total number of the left persons is 423.

**Comment by lodemo** [comment]
Ah no i did not prune images.

In `getLfwPplSorted` the people are returned sorted for number of images, so selecting the first 100 people should result in the same as with the pruned dataset i think?

**Comment by lodemo** [comment]
You can try running my version of the script, see if anything changes.
https://gist.github.com/lodemo/f49ac4a7402d2de3163cf5adfad79d43

**Comment by hardfish82** [comment]
Yes, you are right. Can you share your getData() code with me? I think it may be the key point. By the way, the pretrained model restore() process is very slow(about 10 mins maybe), is there any problem in my envirment?

**Comment by hardfish82** [comment]
Thanks @lodemo . The difference is on the align method. According to [https://cmusatyalab.github.io/openface/demo-3-classifier/](url) , I use dlib align. The former result is based on the dlib_aligned faces. 
I tried the mtcnn align method [https://github.com/davidsandberg/facenet/wiki/Validate-on-LFW](url) just now and got the same result as yours.
It's amazing! Thanks one more time.
![accuracies](https://cloud.githubusercontent.com/assets/3353361/22782168/c50da992-ef00-11e6-82ba-9721f154437f.png)
